### ISSUE TITLE
Support for multiple schemas in postgre

### ISSUE THREAD

**Issue by digoburigo** [drizzle-team/drizzle-kit-mirror#18]
I'm trying to introspect a postgre database with multiple schemas and it's not generating all the tables in all schemas. 

Is there a support for multiple schemas?

**Comment by AndriiSherman** [comment]
Not yet, we were not adding it until someone asks
Added to backlog
Thanks!

**Comment by andresmeidla** [comment]
is there a workaround we could use until this gets implemented?

**Comment by erickreutz** [comment]
Would love to see this also!

**Comment by Yuval-Peled** [comment]
Adding a +1 to this. We organize tables by logical schema (e.g "user" schema, "billing" schema) so the introspection does not work for us at all

**Comment by nicolasey** [comment]
Would be great indeed for modular architecture. 

I was thinking about discovering all schemas using a rule in config > schema variable. 
Instead of direct link to one folder, setup would be for example `./src/modules/**/*.schema.ts` or something similar.

But I do not know if this is good idea for drizzle-kit

**Comment by AndriiSherman** [comment]
I’m currently working on this feature! 
do it’s next to be released in drizzle-kit

**Comment by remorses** [comment]
this is also required to use drizzle with drizzle because the auth users table is created on the `auth` schema

**Comment by anthonyalayo** [comment]
Could we get an update on this?

**Comment by capaj** [comment]
I can hack this with renaming my schema to public for now, but it's not ideal having to do this hack to introspect the DB

**Comment by nobleach** [comment]
I'm not allowed to use the `public` schema. I've tried specifying a different one via the `?search_path=` connection string query param... Doesn't look like it's implemented just yet.

**Comment by anthonyalayo** [comment]
@AndriiSherman could we get an update?

**Comment by AndriiSherman** [comment]
Sorry for not providing updates on this issue for so long. This is the last significant task in the drizzle-kit that requires extensive restructuring due to a few poor decisions regarding the multi-schema architecture. I had been postponing this in favor of implementing a set of different features that required less effort but were important to add.

I'm currently in the process of adding unique constraints, checks, and all fields for indexes to the drizzle kit. After that, I will complete the implementation for multi-schema support in introspect and also address an issue where you can't name a table the same in different schemas

And after that I will be ready to open-source drizzle-kit as well, so you can contribute anything you'd like or just check the implementation

**Comment by AndriiSherman** [comment]
I want to do it at the end of August, but it's always hard to estimate accurately, and I believe that a more appropriate estimate would be in Q4 2023, yes

**Comment by peterferguson** [comment]
Hey didn't want to create a new issue since this is still open but on using the multiple schema introspection there is a small error in the generation. The result looks like this

```ts

export const nextAuth = pgSchema("next_auth");
,export const realtime = pgSchema("realtime");
,export const supabaseFunctions = pgSchema("supabase_functions");
,export const cron = pgSchema("cron");
,export const vault = pgSchema("vault");
,export const extensions = pgSchema("extensions");

``` 

my config looks like

```ts

export default {
	schema: './schema/*',
	out: './migrations',
	driver: 'pg',
	dbCredentials: { connectionString: DB_URL },
	schemaFilter: [
		'public',
		'next_auth',
		'vault',
		'supabase_functions',
		'realtime',
		'extensions',
		'cron',
	],
} satisfies Config
```

**Comment by anthonyalayo** [comment]
@AndriiSherman could we get an update on how this is going along? We're in Q4 now, and multiple schema support is super needed!

**Comment by albertilagan** [comment]
> @AndriiSherman could we get an update on how this is going along? We're in Q4 now, and multiple schema support is super needed!

<img width="823" alt="image" src="https://github.com/drizzle-team/drizzle-kit-mirror/assets/22500561/55305f11-e004-41bf-a759-b3f8da2a133a">

https://discord.com/channels/1043890932593987624/1143797814959292456/1143836511968706622

**Comment by albertilagan** [comment]
> @AndriiSherman could we get an update on how this is going along? We're in Q4 now, and multiple schema support is super needed!

<img width="169" alt="image" src="https://github.com/drizzle-team/drizzle-kit-mirror/assets/22500561/7c1ec273-d079-43a2-8482-ff8e78f28f2d">

**Comment by anthonyalayo** [comment]
Thanks for the find @albertilagan. Looks like there's been no update since this:

> ivanfeliciano — 08/28/2023 6:01 PM
> Is there any timeline for this issue fix?

But what we have so far can unblock me at least.

**Comment by anthonyalayo** [comment]
> And after that I will be ready to open-source drizzle-kit as well, so you can contribute anything you'd like or just check the implementation

If the support is out there, why is drizzle kit still not open source? I see a lot of comments in #kit-discussion about a big desire for it to be open source so that bug fixes can get in.

**Comment by Hebilicious** [comment]
@AndriiSherman 
I wanted to implement a drizzle/kit workflow where on each feature branch, I would create a new pg schema base on the branchname, and run the migrations against that schema with kit.
Then with drizzle, I wanted to be able to switch schema on the fly.

I found a hacky way to do this, which is to manually patch the migrations folder and define my tables with something like this : 

```ts
export const schemaTable = pgSchema(process.env.PG_SCHEMA);
``` 

Unfortunately that currently doesn't work due to the table names being the same in the schema.

Ideally I'd want to not have to define tables with `pgSchema` in drizzle, and simple being able to pass a `defaultPgSearchPath` configuration string to both drizzle-kit *and* drizzle-orm, so that this hacky workaround becomes un-necessary. 

My naive thought process of an implementation would be to automatically inject `SET SEARCH_PATH = example;` right after any operation is sent by drizzle, and to restrict this to a single schema at first.

It's not perfect but it would enable a lot. However there might be some things I'm missing.
Wdyt @AndriiSherman ?

**Comment by AlexBlokh** [comment]
that's fixed as of 0.21.0
